### ISSUE TITLE
winit-0.21 panicks on Wayland

### ISSUE THREAD

**Issue by Abendstolz** [rust-windowing/winit#1441]
Coming from: https://github.com/rust-windowing/winit/issues/1333

While `sway` doesn't crash anymore, I still have no luck running on `wayland` and have to resort to `export WINIT_UNIX_BACKEND=x11`.

I am now using winit-0.21.0 and this is my backtrace captured with RUST_BACKTRACE=full when running on `wayland`:

```
`Running `target/debug/client`                                                                                                                                                               
thread '<unnamed>' panicked at 'Surface was not created with create_surface.', src/libcore/option.rs:1188:5
stack backtrace:
   0:     0x56405b8f0264 - backtrace::backtrace::libunwind::trace::heb43798aede8bd30
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1:     0x56405b8f0264 - backtrace::backtrace::trace_unsynchronized::had2ba7dec4bd2732
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2:     0x56405b8f0264 - std::sys_common::backtrace::_print_fmt::hda61f46e822731b2
                               at src/libstd/sys_common/backtrace.rs:84
   3:     0x56405b8f0264 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hfe37fa5de6572965
                               at src/libstd/sys_common/backtrace.rs:61
   4:     0x56405b917c9c - core::fmt::write::h74887d18db27282c
                               at src/libcore/fmt/mod.rs:1025
   5:     0x56405b8ec527 - std::io::Write::write_fmt::h6808f3d5eceed5e5
                               at src/libstd/io/mod.rs:1426
   6:     0x56405b8f2a5e - std::sys_common::backtrace::_print::hcc0fd4b3552039ef
                               at src/libstd/sys_common/backtrace.rs:65
   7:     0x56405b8f2a5e - std::sys_common::backtrace::print::h1c9c5c1c0505592d
                               at src/libstd/sys_common/backtrace.rs:50
   8:     0x56405b8f2a5e - std::panicking::default_hook::{{closure}}::hefb6085c1ab83a59
                               at src/libstd/panicking.rs:193
   9:     0x56405b8f2751 - std::panicking::default_hook::h1b037d2bf0657ab3
                               at src/libstd/panicking.rs:210
  10:     0x56405b8f313b - std::panicking::rust_panic_with_hook::h787d7f532b084b9a
                               at src/libstd/panicking.rs:471
  11:     0x56405b8f2cee - rust_begin_unwind
                               at src/libstd/panicking.rs:375
  12:     0x56405b913bce - core::panicking::panic_fmt::h76b979c035808e69
                               at src/libcore/panicking.rs:84
  13:     0x56405b913c37 - core::option::expect_failed::h7475be656707bdc0
                               at src/libcore/option.rs:1188
  14:     0x56405b81acce - core::option::Option<T>::expect::h72991860a0fed78b
                               at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libcore/option.rs:348
  15:     0x56405b81acce - smithay_client_toolkit::surface::get_dpi_factor::hb4aa422c8a67bb88
                               at /home/abendstolz/.cargo/registry/src/github.com-1ecc6299db9ec823/smithay-client-toolkit-0.6.4/src/surface.rs:99
  16:     0x56405b7b44e5 - winit::platform_impl::platform::wayland::window::Window::scale_factor::h3186519a375df525
                               at /home/abendstolz/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.21.0/src/platform_impl/linux/wayland/window.rs:276
  17:     0x56405b7b44e5 - winit::platform_impl::platform::wayland::window::Window::inner_size::h1ab64aefd517faab
                               at /home/abendstolz/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.21.0/src/platform_impl/linux/wayland/window.rs:224
  18:     0x56405b4a6b79 - winit::platform_impl::platform::Window::inner_size::h8c5f579ad254ee64
                               at /home/abendstolz/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.21.0/src/platform_impl/linux/mod.rs:276
  19:     0x56405b5f8026 - winit::window::Window::inner_size::h5b3684a0dab3d40a
                               at /home/abendstolz/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.21.0/src/window.rs:478
  20:     0x56405b53405c - client::core::graphics::backends::wgpu::engine::WgpuEngine::get_canvas_size::h68f43805612234f3
                               at src/core/graphics/backends/wgpu/engine.rs:144
  21:     0x56405b663045 - client::game::ecs::executor::Ecs::new::h26732dac76bbd6b4
                               at src/game/ecs/executor.rs:41
  22:     0x56405b65c609 - client::game::states::ingame::Ingame::new::h93944cd27833e8ba
                               at src/game/states/ingame.rs:23
  23:     0x56405b64c3b7 - client::game::states::start_state::h5dc35844a5e50462
                               at src/game/states.rs:18
  24:     0x56405b6490cd - client::single_player::{{closure}}::hb5e2fb8163c0b4bf
                               at src/lib.rs:49
  25:     0x56405b5c1c63 - std::sys_common::backtrace::__rust_begin_short_backtrace::hb88f138eebc7e8f5
                               at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/sys_common/backtrace.rs:136
  26:     0x56405b357cc4 - std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}::haffe1788c9d9c6cc
                               at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/thread/mod.rs:469
  27:     0x56405b509a66 - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hdd76686aeb6836b5
                               at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/panic.rs:318
  28:     0x56405b4641c3 - std::panicking::try::do_call::h6265eced88546639
                               at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/panicking.rs:292
  29:     0x56405b8fbafa - __rust_maybe_catch_panic
                               at src/libpanic_unwind/lib.rs:78
  30:     0x56405b463368 - std::panicking::try::hbc4e5c77f5193509
                               at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/panicking.rs:270
  31:     0x56405b509c46 - std::panic::catch_unwind::h614e989c3dcbe12e
                               at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/panic.rs:394
  32:     0x56405b356f9d - std::thread::Builder::spawn_unchecked::{{closure}}::h551128c91cee7eef
                               at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/thread/mod.rs:468
  33:     0x56405b2b2bb4 - core::ops::function::FnOnce::call_once{{vtable.shim}}::haabc5b55b0e0cd4a
                               at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libcore/ops/function.rs:232
  34:     0x56405b8e44cf - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::h7bfd6e98bcf6c67e
                               at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/liballoc/boxed.rs:1022
  35:     0x56405b8fabc0 - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::h9979661149bbc525
                               at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/liballoc/boxed.rs:1022
  36:     0x56405b8fabc0 - std::sys_common::thread::start_thread::had0bb4166c065276
                               at src/libstd/sys_common/thread.rs:13
  37:     0x56405b8fabc0 - std::sys::unix::thread::Thread::new::thread_start::hbc75223688f972da
                               at src/libstd/sys/unix/thread.rs:80
  38:     0x7f56b8db04cf - start_thread
  39:     0x7f56b8cc52d3 - clone
  40:                0x0 - <unknown>


```

**Comment by kchibisov** [comment]
Are you doing some multi-threading? The panic is on `Mutex` AFAICS.  Anyway, more context could be nice. Do you have the app to share?

**Comment by Abendstolz** [comment]
I use `specs` and also one more thread where the `winit-event-loop` resides (which then pumps the interesting events to the `ecs` via a `channel`)

I can share the whole code, but it's in a private repository on my own server and not on `github`. I'd rather not publicly share it at the moment, so if we can find a way that suits you, sure!

**Comment by kchibisov** [comment]
Oh, nvm. I'm a bit blind. You're getting `Surface was not created with create_surface.` So somehow a wrong surface reached winit or something.

>can share the whole code, but it's in a private repository on my own server and not on github. I'd rather not publicly share it at the moment, so if we can find a way that suits you, sure!

I'll investigate first in winit with your backtrace.

**Comment by vberger** [comment]
@Abendstolz are you using some special way to setup the winit Window? From the backtrace, it looks like the main wayland `wl_surface` that winit is trying to use was not initialized by it.

**Comment by Abendstolz** [comment]
@vberger Not that I am aware of :)

```
    pub fn new(settings: &dyn GfxSettings) -> Result<(Self, impl EventHandler), CreationError> {
        let event_loop = EventLoop::new();
        let size = match settings.get_surface_dimensions() {
            Some((width, height)) => LogicalSize::new(width, height),
            None => {
                info!("Settings don't contain surface dimensions - best guess in effect");
                let primary_monitor = event_loop.primary_monitor();
                let size = primary_monitor.size();
                let scale_factor = primary_monitor.scale_factor();
                LogicalSize::from_physical(size, scale_factor)
            }
        };

        info!("Surface size: {:?}", size);

        let adapter = Adapter::request(
            &RequestAdapterOptions {
                power_preference: PowerPreference::HighPerformance,
            },
            BackendBit::all(),
        )
        .ok_or(CreationError::Gpu("No Adapter/GPU was found".to_owned()))?;

        let (device, mut queue) = adapter.request_device(&DeviceDescriptor {
            extensions: Extensions {
                anisotropic_filtering: settings.is_anisotropic_filtering_enabled(),
            },
            limits: Limits::default(),
        });

        let window = WindowBuilder::new()
            .with_title("NerdHunt")
            .with_inner_size(size)
            .with_maximized(true)
            .build(&event_loop)
            .map_err(|e| CreationError::Window(format!("Could not create window: {:?}", e)))?;

        let surface = Surface::create(&window);

        let swap_chain_descriptor = Self::build_swap_chain_descriptor(size.width, size.height);

        let swap_chain = device.create_swap_chain(&surface, &swap_chain_descriptor);

        let depth_texture = depth_texture(&device, &swap_chain_descriptor).create_default_view();
        let terrain_renderer = TerrainRenderer::new(&device, &swap_chain_descriptor, &mut queue);
        let entity_renderer: EntityRenderer<EntityId> = EntityRenderer::new();

        let moi = Self {
            device,
            queue,
            window,
            surface,
            swap_chain_descriptor,
            swap_chain,

            terrain_renderer,
            entity_renderer,
            depth_texture,
        };
        let eh = WinitEventHandler::new(event_loop);
        Ok((moi, eh))
    }
```

Also note, that the window itself gets created, more precisely `new` in `Window`  `platform_impl/linux/wayland/window.rs` reaches it's end and returns an `Ok(...)`

**Comment by vberger** [comment]
> That that point in time I can't see how the threading creates an issue, my wrapper around all things winit and wgpu is still on the same thread it was created on and the ECS receives the wrapper by value and just tries to get the (width, height) of the canvas to initialize a Camera.

Hmm, wait, so you are calling the `inner_size()` method of `Window` from an other thread than the one on which the `EventQueue` was created, right? I think I got the issue.

**Comment by Abendstolz** [comment]
If the `EventQueue` refers to the `EventLoop` then yes.

Does that mean this behaviour is simply disallowed? I've fixed it on my side by simple not getting the window dimensions that way.

**Comment by vberger** [comment]
No, this just confirms that this is a bug in SCTK, as I expected. This will be fixed in the next point-release of the lib (on which winit depends). I'll notify this issue when it is, at which point you'll just need to `cargo update`.

**Comment by Abendstolz** [comment]
Cool! If I can be of any more assistance, just shoot.
Thanks for your time and help!

**Comment by vberger** [comment]
I just released SCTK version 0.6.5, which contains the fix. Can you run `cargo update` and confirm if it indeed solves the issue?

**Comment by Abendstolz** [comment]
@vberger I reintroduced a call to `window.inner_size()` in my pipeline at the same position it was before and log the result.

No crash! Woohoo!

Thanks a bunch and have a wonderful weekend :)
